codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt

```diff
@@ -0,0 +1,642 @@
+    codegenContext: CodegenContext,
+    protocol: Protocol,
+) : ServerProtocolGenerator(
+    codegenContext,
+    protocol,
+    MakeOperationGenerator(codegenContext, protocol, HttpBoundProtocolBodyGenerator(codegenContext, protocol)),
+    ServerHttpProtocolImplGenerator(codegenContext, protocol),
+)
+
+/*
+ * Class used to expose Rust server traits types. Is is used in ServerHttpProtocolGenerator and ServerProtocolTestGenerator.
+ */
+class HttpServerTraits {
+    fun parseHttpRequest(runtimeConfig: RuntimeConfig) = RuntimeType(
+        "ParseHttpRequest",
+        dependency = CargoDependency.SmithyHttpServer(runtimeConfig),
+        namespace = "aws_smithy_http_server::request"
+    )
+
+    fun serializeHttpResponse(runtimeConfig: RuntimeConfig) = RuntimeType(
+        "SerializeHttpResponse",
+        dependency = CargoDependency.SmithyHttpServer(runtimeConfig),
+        namespace = "aws_smithy_http_server::response"
+    )
+
+    fun serializeHttpError(runtimeConfig: RuntimeConfig) = RuntimeType(
+        "SerializeHttpError",
+        dependency = CargoDependency.SmithyHttpServer(runtimeConfig),
+        namespace = "aws_smithy_http_server::response"
+    )
+}
+
+/*
+ * Generate all operation input parsers and output serializers for streaming and
+ * non-straming types.
+ */
+private class ServerHttpProtocolImplGenerator(
+    private val codegenContext: CodegenContext,
+    private val protocol: Protocol,
+) : ProtocolTraitImplGenerator {
+    private val logger = Logger.getLogger(javaClass.name)
+    private val symbolProvider = codegenContext.symbolProvider
+    private val model = codegenContext.model
+    private val errorType = RuntimeType("error", null, "crate")
+    private val runtimeConfig = codegenContext.runtimeConfig
+    private val httpBindingResolver = protocol.httpBindingResolver
+    private val operationDeserModule = RustModule.private("operation_deser")
+    private val operationSerModule = RustModule.private("operation_ser")
+    private val smithyJson = CargoDependency.smithyJson(runtimeConfig).asType()
+    private val httpServerTraits = HttpServerTraits()
+
+    private val codegenScope = arrayOf(
+        "ParseHttpRequest" to httpServerTraits.parseHttpRequest(runtimeConfig),
+        "ParseStrictResponse" to RuntimeType.parseStrictResponse(runtimeConfig),
+        "SerializeHttpResponse" to httpServerTraits.serializeHttpResponse(runtimeConfig),
+        "SerializeHttpError" to httpServerTraits.serializeHttpError(runtimeConfig),
+        "JsonObjectWriter" to smithyJson.member("serialize::JsonObjectWriter"),
+        "Http" to RuntimeType.http,
+        "Bytes" to RuntimeType.Bytes,
+        "Error" to errorType.member("Error"),
+        "LazyStatic" to CargoDependency("lazy_static", CratesIo("1.4")).asType(),
+        "Regex" to CargoDependency("regex", CratesIo("1.0")).asType(),
+        "PercentEncoding" to CargoDependency("percent-encoding", CratesIo("2.1.0")).asType(),
```

Comment:
I would probably move these to `CargoDependency`. That way, it's obvious what version of each crate we're using and we don't end up with version mismatches

Reply:
Makes sense, I'll do it.